File: bae-avalonia/Stores/ArtworkLoadingStore.cs
using System;
using uniffi.bae_bridge;

namespace Bae.Desktop;

/// <summary>The retained post-open CacheEager artwork operation.</summary>
internal sealed class ArtworkLoadingStore
{
    private readonly Action _cancel;

    public ArtworkLoadingStore(Action cancel)
    {
        _cancel = cancel;
    }

    public BridgeEagerCacheFillStatus Status { get; private set; }
        = new BridgeEagerCacheFillStatus.NotRunning();

    public event Action? Changed;

    public void Apply(BridgeEagerCacheFillStatus status)
    {
        Status = status;
        Changed?.Invoke();
    }

    public void Cancel() => _cancel();
}

File: BaeKit/Sources/BaeKit/Services/ArtworkLoadingStore.swift
import Foundation
import Observation

/// The post-open CacheEager artwork operation for one library. The bridge is
/// its sole writer; views read the retained status and may request cancellation.
@MainActor
@Observable
public final class ArtworkLoadingStore {
    public private(set) var status: BridgeEagerCacheFillStatus = .notRunning
    private let cancelAction: @Sendable () -> Void

    public init(cancel: @escaping @Sendable () -> Void) {
        self.cancelAction = cancel
    }

    public func apply(_ status: BridgeEagerCacheFillStatus) {
        self.status = status
    }

    public func cancel() {
        cancelAction()
    }
}
